src/lib/spinel/radio_spinel.cpp

```diff
@@ -0,0 +1,3140 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the spinel based radio transceiver.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdarg.h>
+#include <stdlib.h>
+
+#include <openthread/dataset.h>
+#include <openthread/logging.h>
+#include <openthread/platform/diag.h>
+#include <openthread/platform/time.h>
+
+#include "common/code_utils.hpp"
+#include "common/encoding.hpp"
+#include "common/instance.hpp"
+#include "common/new.hpp"
+#include "common/settings.hpp"
+#include "common/settings_driver.hpp"
+#include "lib/platform/exit_code.h"
+#include "lib/spinel/radio_spinel.hpp"
+#include "lib/spinel/spinel.h"
+#include "lib/spinel/spinel_decoder.hpp"
+#include "meshcop/dataset.hpp"
+#include "meshcop/meshcop_tlvs.hpp"
+#include "radio/radio.hpp"
+#include "thread/key_manager.hpp"
+
+#ifndef MS_PER_S
```

Comment:
@zhanglongxia I noticed you added new `radio_spinele.cpp` (which is not using CRP) but it does not seem to be  added in any of makefiles,  and we still keep the template `radio_spinel_impl.hpp`? 

Should we change them as well? Or not add `radio_spinel.cpp` in this PR (add in follow up where CRTP model is replaced?)

Reply:
Sorry, I pushed a wrong branch to this commit. I have corrected the code. Please review it again.

Reply:
No worries. 😃  LGTM.